File: Reverse Linked List.java
Given the head of a singly linked list, reverse the list, and return the reversed list.
  
Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]

class Solution 
{
    public ListNode reverseList(ListNode head) 
    {      
        // Edge case where the list is empty.
        if(head == null) return null;
        
        // Two helper pointers move at the same time.
        ListNode prev = null;
        ListNode nextNode = head.next;
        
        // While the nextNode has not reached the end of the list.   
        // Reverse head.next. Have it pointing to it's previous node.  ----------> Line 22.
        // Advance all three pointers.                                 ----------> Line 23 24 25.
        while(nextNode != null)
        {
            head.next = prev;
            prev = head;
            head = nextNode;
            nextNode = nextNode.next;
        }
        
        // At the very last iteration of the loop, head would be at the last node in the list.
        // Because the condition is false, line 22 would not get executed. 
        head.next = prev;
        return head;
    }
}

Before 1st iteration:            null        1     ->       2       ->      3      ->     4  
                                 prev       head         nextNode


1st iteration:                   null   <-     1            2       ->      3      ->     4   
                                  prev       head        nextNode  

                                 null   <-     1            2       ->      3      ->     4   
                                          prev,head        nextNode  

                                 null   <-     1            2       ->      3      ->     4    
                                             prev        head,nextNode  
                                            
                                 null   <-     1            2       ->      3      ->     4    
                                             prev        head           nextNode  


2nd iteration:                    null   <-     1      <-      2            3      ->     4    
                                             prev           head          nextNode  
                                             
                                  null   <-     1      <-      2            3      ->     4    
                                                           prev,head      nextNode  
                                                           
                                  null   <-     1      <-      2            3      ->     4    
                                                              prev      head,nextNode 
                                                              
                                  null   <-     1      <-      2            3      ->     4    
                                                              prev      head          nextNode  


                                             
3rd iteration:                    null   <-     1      <-      2      <-      3           4    
                                                              prev          head       nextNode  
              
                                  null   <-     1      <-      2      <-      3           4    
                                                                          prev, head   nextNode  
                                                                          
                                  null   <-     1      <-      2      <-      3           4    
                                                                            prev      head,nextNode  

                                  null   <-     1      <-      2      <-      3           4           null
                                                                            prev         head       nextNode
    
    
 Condition failed                 null   <-     1      <-      2      <-      3     <-    4
                                                                            prev         head        
  

File: Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        # 3 pointers approach
        prev, new_head, next_node = None, head, head.next
        while next_node:
            new_head.next = prev
            prev = new_head
            new_head = next_node
            
            next_node = next_node.next
        
        new_head.next = prev
        return new_head
----------------------------------------------------------------------------------------
 class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        prev = None
        
        while head:
            next_node = head.next
            head.next = prev
            prev = head
            head = next_node
        
        return prev
 ------------------------------------------------------------------------------------------
 class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        
        return new_head
